src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,147 @@
+    private int size;
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        //here we can't use method checkElementIndex(index)
+        // because we add +1 to size
+        //useless to create new method only for this case
+        Objects.checkIndex(index, size + 1);
+        if (size == index) {
+            add(value);
+        } else {
+            Node<T> nodeByIndex = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            newNode.next.prev = newNode;
+            if (index != 0) {
+                newNode.prev.next = newNode;
+            } else {
+                head = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldVal = node.element;
+        node.element = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        T removedElement;
+        if (index == 0) {
+            removedElement = head.element;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            removedElement = prev.next.element;
+            prev.next = prev.next.next;
+            if (index == size - 1) {
+                tail = prev;
+            }
+        }
+        size--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        int index = 0;
+        Node<T> current = head;
+        while (current != null) {
+            if (object == current.element || object != null && object.equals(current.element)) {
+                remove(index);
+                return true;
+            }
+            current = current.next;
+            index++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
```

Comment:
we can improve performance here
if the index is in the first part of the list, you can start the movement from the beginning, otherwise you can start the movement from the end